WIP: Discrete and synchronized targets support

sliwy opened the pull request.

MFattouh commented:
I see a number of issues for cropped decoding with the current implementation, which uses targets as anchor (stop) points and then slices a window with the specified size before the target.

- [x] Depending on the specified window size, we could end up with windows that span a number of targets. However, currently, only the last target is returned. This is the expected behavior for Trial-wise decoding, but not in cropped decoding. We can control this behavior by adding a new optional parameter to WindowsDataset to return all targets within the window or just the last one.

- [ ] The new create_windows_from_target_channels windowing function implemented by this PR would create overlapping windows with duplicate targets, which is not desired, because targets are assumed to appear only once in a single dataset iteration. However, we can use the old windowing functions from braindecode to iterate over the entire trial, but we want to emphasize the difference between trials-wise and cropped decoding windowing functions in the documentation and the tutorial.

- [ ] Proper validation that accounts for the receptive field of the models by considering fixed number of targets (specified by the user) per trial for validation.

robintibor commented:
sounds like you are thinking about the important issues, removing duplicate targets etc can always be a bit tricky, but we do have quite some logic there already in the scoring.py code, you can also take inspiration from it when you write some new code to tackle your use case

robintibor commented:
So summary what I recall now from what I said.
* We can "create fixed length windows", if targets are simply inside the compute window, they are matching, we may just have to remove some at beginning if network predicts less
* If we leave NaNs inside returned targets from dataset, matching afterwards could be quite easy: creating a mask that removes NaNs on the targets and applying the same mask to the predictions should get the correct matching already
* We could make a generic loss wrapper that accepts predictions, targets and a loss function, removes the values at NaN indices as described above and then applies the loss function to the remaining values

robintibor commented:
Please make sure that the naming is clear, like I just had some quick glance and the naming with ECoG seems a bit confusing to me. Keep in mind target-timeseries decoding is not directly tied to ECoG, you could do target-timeseries decoding directly from EEG as well... so make sure you don't introduce confusing naming there... Also make sure it is clear what this still has to do with BCI Competition etc., that it is clearly described.

sliwy commented on braindecode/datasets/ecog_bci_competition.py:
```diff
@@ -0,0 +1,113 @@
+    """
+    possible_subjects = [1, 2, 3]
+
+    def __init__(self, path, subject_ids=None):
```
@robintibor @agramfort For now to create this dataset user must specify path to the data (folder with .mat files downloaded from BCI IV competition site). To download the data you have to sign in with your email and password received in an email. Do you see any easier option to do that (without additional steps outside braindecode)? Do we want to create any specific path where we can put this kind of data (like mne_data)?

robintibor replied:
so this dataset is not in moabb? I mean if you have to manually download anyways I feel supplying path seems fine to me...

sliwy replied:
This dataset is not in moabb, so probably we will keep the dataset path as a parameter

sliwy replied:
This dataset will be stored in `mne_data` as MOABB datasets, so problem solved :)

robintibor commented on examples/plot_bcic_iv_4_ecog_trial.py:
```diff
@@ -0,0 +1,287 @@
+#
+from sklearn.metrics import r2_score
+
+DATASET_PATH = '/home/maciej/projects/braindecode/BCICIV_4_mat'
```
I assume you may have to mock the dataset like Lukas did in https://github.com/braindecode/braindecode/pull/276 for this to run on CI eventually? Or what is your plan?

sliwy replied:
Yeah, I will try to prepare something similar! :)

sliwy replied:
This one is still to be done. Mocking and testing BCICompetitionDataset4

agramfort commented on braindecode/datasets/ecog_bci_competition.py:
```diff
@@ -0,0 +1,113 @@
+    http://www.bbci.de/competition/iv/results/ds4/true_labels.zip
+
+    Data can be downloaded from (Dataset 4):
+    http://www.bbci.de/competition/iv/#dataset4
```
these data are super old. I am sure we could now make them available without password eg storing them on OSF.

I would contact the authors of the competition to ask for permission.

Maybe @cbrnr nows who to ask this.

I really don't like that an example does not work out of the box. And you won't be able to test it in CI or to build any doc. It's for me a blocker.

cbrnr replied:
Not sure who "owns" the data, but if it's BBCI I'd contact Benjamin Blankertz. If not, either Gerwin Schalk or Kai J. Miller are listed. I know Benjamin and Gerwin, if you want I could ask them.

agramfort commented:
let's talk about the design of this one on discord when you have time.

agramfort commented:
can you send an email to Benjamin and Gerwin with us in CC to explain what
we do here?

thx !

>

cbrnr commented:
Sure, will do!

agramfort commented:
@sliwy the fetcher should pull the data from https://searchworks.stanford.edu/view/zk881ps0522
no registration required.

sliwy commented:
@agramfort I've seen the email, it's perfect. For the implementation, should we create `braindecode_data` folder to save the data like in mne or can we use `mne_data` folder like moabb even if it's not meant for data from mne? 

@cbrnr Thank you for  your help! :)

robintibor commented:
Please let me also know if you need any support/have any questions at the moment with this @MFattouh @sliwy :)

sliwy commented:
@robintibor that's a good question. Before the weekend I started working on loading the dataset from the repository mentioned above. Currently I'm on vacations and I didn't have enough time to work together with @MFattouh. Next week should be a little bit easier than last weeks, so I hope we can make it. :)

sliwy commented:
A quick update on the PR, I asked Gerwin Schalk about another possible solution to have this dataset openly available. From the link provided by them we can download only a bigger dataset that was probably used to prepare the data. I f we want to use that data repository it would need from us some additional steps of data extraction, downloading additional not needed data and more information from dataset providers.

Instead I asked Gerwin Schalk if we can just take the files from BCI Competition and put them on OSF (as @agramfort suggested). We will give credits to the owners of the dataset and with much smaller effort we will have the dataset openly available. Still there are some code changes to be done in the meantime finish the PR.

robintibor commented:
so how it's going any support needed @sliwy @MFattouh ?

sliwy commented:
Great that you ask, so for now a few things are clear:
1. I had some discussions with Kai J Miller about the dataset and its availability. Two days ago the dataset was uploaded to the ECoG library, so we finally have an easy access to the dataset. I have already modified the dataset downloading part to use new ECoG library instead of path to earlier downloaded files. This should be fine.
2. I think that trialwise example is almost ready. I added some tests yesterday, I'll take one more look at it (to add some more tests) this week.
3. For the cropped decoding things are more complicated but still quite close to the end. I think that we need some improvement in the last part of the tutorial where there are some big code snippets with for-loops etc. It can be placed inside functions but I feel that @MFattouh can have more to explain that. Of course @robintibor you can take a look at cropped decoding example as you're the master of cropped decoding. :)
4. @robintibor do you know why there is no documentation built based on this PR? I didn't have time to check that but I see that there is only one check done here.

robintibor commented:
Maybe the whole draft PR logic prevents CI from running? I have remove this draft stuff. Maybe still a commit needed  to make it run? Or needs rebase...

robintibor commented:
Well might anyways be a good step to try to rebase @sliwy ? Then things may run?

robintibor commented on examples/plot_bcic_iv_4_ecog_cropped.py:
```diff
@@ -0,0 +1,358 @@
+######################################################################
+# First, we load the data. In this tutorial, we use the functionality of
+# braindecode to load datasets through
+# `MOABB <https://github.com/NeuroTechX/moabb>`__ to load the BCI
```
This text and also a lot of other text in this example is not correct/still fits to old example. Feel free to also shorten text quite a bit, but at least the text that is there should be correct.

robintibor commented:
I made some small comments in the cropped example @sliwy. Also for final for loops, could https://github.com/braindecode/braindecode/blob/2dbd40cc7951934ad32892770536e360b6188920/braindecode/training/scoring.py#L303 help there? Otherwise what is the status, are there things still open that you want to do? @sliwy @MFattouh

sliwy commented:
@robintibor this `predict_trials` function may be really helpful to make cropped decoding example more accessible. Refactor of the cropped decoding example still on my list. :)

robintibor commented:
@sliwy feel free to adapt the predict_trials function to be more flexible if that is causing the error

sliwy commented on examples/plot_bcic_iv_4_ecog_cropped.py:
```diff
@@ -0,0 +1,381 @@
+                                        test_size=0.2,
+                                        shuffle=False)
+valid_set = test_set
+# valid_set = torch.utils.data.Subset(train_set, idx_valid)
```
@robintibor is there any way to subset a cropped dataset without raising error while validating that a windows indices have to start from zero?

robintibor replied:
hmhm very good question. I guess no and we maybe should implement it? I think usually our expectation was that splitting is mostly done before windowing. However, I have also split valid datasets off windows/cropped datasets, but there usually I knew how many windows there were per trial and could ensure by hand that everything will be correct.. probably we should have some function for it? But to not delay this PR further is it possible to move the splitting to before the windowing?

robintibor commented on examples/plot_bcic_iv_4_ecog_cropped.py:
```diff
@@ -0,0 +1,381 @@
+fig, ax1 = plt.subplots(figsize=(8, 3))
+ax1.plot(x_time, y_test[0, :], label='Target')
+ax1.plot(x_time, preds_test[0, :], label='Predicted')
+ax1.xlabel('Time [s]')
+ax1.ylabel('Finger flexion')
```
```suggestion
ax1.set_xlabel('Time [s]')
ax1.set_ylabel('Finger flexion')
```

needs to be corrected @sliwy

sliwy commented:
@robintibor Two things that left to be done:
1. Improve tutorial appearance (sections, spacing etc.)
2. Change splitting in the cropped decoding (split on raw datasets).

Probably this weekend will be enough to finish. :)

robintibor commented:
Great would be amazing @sliwy ! Then we can move towards next release

robintibor commented on braindecode/datasets/bcicomp.py:
```diff
@@ -0,0 +1,156 @@
+        -------
+
+        """
+        sign = 'BCICompetitionIVDataset4'
```
sign?

robintibor commented on examples/plot_bcic_iv_4_ecog_trial.py:
```diff
@@ -0,0 +1,389 @@
+    batch_size=batch_size,
+    callbacks=[
+        'r2',
+        ('valid_pearson_r', EpochScoring(pearson_r_score, lower_is_better=False,
```
```suggestion
        ('valid_pearson_r', EpochScoring(pearson_r_score, lower_is_better=False, on_train=False,
```

more symmetric

robintibor commented:
so @sliwy how is going? so this is ready to merge for you? did you see small comments from me?

sliwy commented:
@robintibor I was a little bit busy, from my side it's ready to merge, I fixed the comments from you, thanks :)

robintibor commented on braindecode/training/scoring.py:
```diff
@@ -206,14 +207,96 @@ def on_epoch_end(self, net, dataset_train, dataset_valid, **kwargs):
+            mask = ~np.isnan(targets)
+            # select valid targets that have a matching predictions
+            masked_targets = targets[mask]
+            if mask.shape[0] == 1:
```
could you put a comment here when/why this case happens? @sliwy

robintibor commented:
Great work @sliwy! I had a look through now, could you add the comment, in that position I was not sure why which case appears, and would be great to have it documented there :) 
After, I would merge. then we could do  a new braindecode release soon after ICLR deadline next tuesday!

In the future we may wanna think about adding a 'target_channels' argument  to the `create_windows_from_target_channels` to ease usability using that API (keeping WindowsDataset same with purely `targets_from` logic). But for now seems all fine to me!

robintibor commented on braindecode/datasets/bcicomp.py:
```diff
@@ -91,18 +93,18 @@ def download(path=None, force_update=False, verbose=None):
-        destination = _url_to_local_path(DATASET_URL, osp.join(path, key_dest))[
-                      :-4] + '/BCI_Competion4_dataset4_data_fingerflexions'
-        if len(list(glob.glob(destination + '/*.mat'))) == 6:
+        # We do not use mne _url_to_local_path due to ':' in the url that causes problems on Windows
```
interesting, maybe also interesting to know for @agramfort ?

robintibor commented:
Amazing amazing @sliwy merged
